### PR TITLE
Batch inference documentation

### PR DESCRIPTION
**Pull request by vlad-tokarev**
for https://github.com/odahu/odahu-flow/issues/500
image

### CODE REVIEW

File: source/architecture_batch.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+using ML Model
+
+This dataset could have relatively large size and using blocking web api does not fit this usecase perfectly
+
+So we should provide async API that provide way to specify 
+
+1. Where input data is located (object storage link: s3, gcs, azure, etc)
+2. Where model is located (object storage link: s3, gcs, azure, etc) OR maybe some specific ML Registry such as MLFlow registry
+3. Where output data should be uploaded (object storage link: s3, gcs, azure, etc)
+4. Triggers that allow to run this job
```
Comment (BPylypenko) [inline source/architecture_batch.md]: small grammar fixes
``` suggestion
A user wants to run workloads using a trigger (cron, webhook, cloud events) that makes predictions on a dataset using ML Model.

This dataset could have a relatively large size and the use of blocking web API does not fit this use case perfectly.

So we should provide an async API that provides a way to specify: 

1. Where input data is located (object storage link: s3, gcs, azure, etc)
2. Where the model is located (object storage link: s3, gcs, azure, etc) OR maybe some specific ML Registry such as MLFlow registry
3. Where output data should be uploaded (object storage link: s3, gcs, azure, etc)
4. Triggers that allow running this job.
```

File: source/architecture_batch.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+- Spec.modelPath must be directory in remote storage (not file) and must contain
+  `odahuflow.model.yaml` file which should have next properties
+  ```yaml
+  modelName: <Model name>
+  modelVersion: 1.3
+  ```
+  and any other files, for example model configuration that can be loaded to model in inference script
+  
+- Spec.inputPath must be a file in remote storage (not directory) and be one of next two options:
```
Comment (BPylypenko) [inline source/architecture_batch.md]:
```suggestion
- Spec.inputPath must be a file in remote storage (not a directory) and be one of the next two options:
```

File: source/architecture_batch.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  - option 1: json file that contains data saved in `Inference Request` object 
+    according to [Predict Protocol - Version 2](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#inference-request-json-object)
+  - option 2: .tar.gz file that contain a compressed bundle of json files the same format as json from option 1
```
Comment (BPylypenko) [inline source/architecture_batch.md]:
```suggestion
  - option 2: .tar.gz file that contains a compressed bundle of json files in the same format as json from option 1
```

File: source/architecture_batch.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+- Spec.outputPath must be a file in remote storage (not directory) and be one of next two options:
```
Comment (BPylypenko) [inline source/architecture_batch.md]:
```suggestion
- Spec.outputPath must be a file in remote storage (not a directory) and be one of the next two options:
```

File: source/architecture_batch.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  - option 1: json file that contains data saved in `Inference Response` object 
+    according to [Predict Protocol - Version 2](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#inference-response-json-object)
+  - option 2: .tar.gz file that contain a compressed bundle of json files the same format as json from option 1
```
Comment (BPylypenko) [inline source/architecture_batch.md]:
```suggestion
  - option 2: .tar.gz file that contain a compressed bundle of json files in the same format as json from option 1
```

File: source/architecture_batch.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  outputConnection: output-data
+  modelConnection: model-data
+  # If data will be not available in the moment of launch the predictor 
+  # will wait some time (retries * delayBetweenRetries) before stop with exception
+  waitForInputData:
+    retries: 15  
+    delayBetweenRetries: 120 
+  triggers:
+    webhook:
+      enabled: true
```
Comment (BPylypenko) [inline source/architecture_batch.md]: We plan, for now, to add only `webhook` triggers or, maybe, it worth creating a separate entity for triggers? To have the possibility update only trigger URI or type of trigger?

**Review by BPylypenko — CHANGES_REQUESTED** [review]

**Review by keshamin — APPROVED** [review]

**Review by BPylypenko — APPROVED** [review]

**Review by easokol — APPROVED** [review]

**Review by karbyshevds — APPROVED** [review]

File: source/ref_batch.rst [hunk 7]
```diff
@@ -0,0 +1,124 @@
+
+###################
+Batch Inference
+###################
+
+This section describes API and protocols related to Batch inference using ODAHU
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
This section describes API and protocols related to Batch inference using ODAHU.
```

File: source/ref_batch.rst [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+ODAHU Batch Inference feature allows to user get inferences using model for large datasets that are delivered not via
+WEB API but other mechanisms.
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
ODAHU Batch Inference feature allows user to get inferences using ML model for large datasets that are delivered asyncronously, not via
HTTP API, but through other mechanisms.
```

File: source/ref_batch.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+Currently Batch Inference supports next ways to delivery data for forecasting:
+
+- Object storage
+
+  - GCS
+  - S3
+  - Azureblob
+
+For the future we consider to add ability to process data directly from Kafka topic and other async data sources.
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
In future we consider to add ability to process data directly from Kafka topic and other async data sources.
```

File: source/ref_batch.rst [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+InferenceService
+=================
+
+InferenceService represents next required entities:
+
+- Predictor docker image that contains predictor code
+- Model files location on object storage (directory or .zip / .tar.gz archive)
+- Command and Arguments that describe how to execute image
+
+When user trained a model then he should build and image with code that follows `Predictor code protocol`_ and register
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
When a user trains a model then they should build an image with code that follows `Predictor code protocol`_ and register
```

File: source/ref_batch.rst [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+.. openapi:: odahu-core-openapi.yaml
+   :paths:
+      /api/v1/batch/service
+
+
+==================
+InferenceJob
+==================
+
+``InferenceJob`` describes forecast process that was triggered by on of the triggers of ``InferenceService``.
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
``InferenceJob`` describes forecast process that was triggered by one of the triggers in ``InferenceService``.
```

File: source/ref_batch.rst [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+If ``[].spec.triggers.webhook`` is enabled then its possible to run ``InferenceJob`` by making POST request as described
+below. Be default webhook trigger is enabled. Note, that currently its the only one way to trigger jobs.
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
below. By default webhook trigger is enabled. Note, that currently its the only one way to trigger jobs.
```

File: source/ref_batch.rst [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+   :paths:
+      /api/v1/batch/job
+
+
+
+*********************************
+Predictor code protocol
+*********************************
+
+ODAHU Platform launches docker image that provided by user as ``[].spec.image`` (InferenceService_) and guaranties the
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
ODAHU Platform launches docker image provided by user as ``[].spec.image`` (InferenceService_) and guarantees the
```

File: source/ref_batch.rst [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+**************
+API Reference
+**************
+
+
+=================
+InferenceService
+=================
+
+InferenceService represents next required entities:
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
InferenceService represents the following required entities:
```

File: source/ref_batch.rst [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+###################
+Batch Inference
+###################
+
+This section describes API and protocols related to Batch inference using ODAHU
+
+ODAHU Batch Inference feature allows to user get inferences using model for large datasets that are delivered not via
+WEB API but other mechanisms.
+
+Currently Batch Inference supports next ways to delivery data for forecasting:
```
Comment (keshamin) [inline source/ref_batch.rst]:
```suggestion
Currently Batch Inference supports the following ways to delivery data for forecasting:
```

File: source/ref_batch.rst [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+=================
+InferenceService
+=================
+
+InferenceService represents next required entities:
+
+- Predictor docker image that contains predictor code
+- Model files location on object storage (directory or .zip / .tar.gz archive)
+- Command and Arguments that describe how to execute image
```
Comment (keshamin) [inline source/ref_batch.rst:38]: Isn't it also responsible for input/output data location?

**Review by keshamin — APPROVED** [review]

**Review by easokol — APPROVED** [review]

**Review by karbyshevds — APPROVED** [review]

**Pull request merged**